network/network.go

```diff
@@ -0,0 +1,182 @@
+	}
+	if config.KubernetesClient == nil {
+		return nil, microerror.Maskf(exporterkit.InvalidConfigError, "%T.KubernetesClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(exporterkit.InvalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.Host == "" {
+		return nil, microerror.Maskf(exporterkit.InvalidConfigError, "%T.Host must not be empty", config)
+	}
+	if config.Namespace == "" {
+		return nil, microerror.Maskf(exporterkit.InvalidConfigError, "%T.Namespace must not be empty", config)
+	}
+	if config.Port == "" {
+		return nil, microerror.Maskf(exporterkit.InvalidConfigError, "%T.Port must not be empty", config)
+	}
+	if config.ServiceName == "" {
+		return nil, microerror.Maskf(exporterkit.InvalidConfigError, "%T.ServiceName must not be empty", config)
+	}
+
+	networkCollector := NetworkCollector{
+		dialer:           config.Dialer,
+		kubernetesClient: config.KubernetesClient,
+		logger:           config.Logger,
+
+		host:        config.Host,
+		namespace:   config.Namespace,
+		port:        config.Port,
+		serviceName: config.ServiceName,
+
+		count:      map[string]float64{},
+		errorCount: map[string]float64{},
+
+		total: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "", "check_total"),
+			"Total number of network checks.",
+			[]string{"host"},
+			nil,
+		),
+		errorTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "", "check_error_total"),
+			"Total number of network check errors.",
+			[]string{"host"},
+			nil,
+		),
+		latency: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "", "check_seconds"),
+			"Time taken to successfully check network.",
+			[]string{"host"},
+			nil,
+		),
+		kubernetesErrorTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "", "kubernetes_error_total"),
+			"Total number of errors reaching Kubernetes API.",
+			nil,
+			nil,
+		),
+	}
+
+	return &networkCollector, nil
+}
+
+func (c *NetworkCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.total
+	ch <- c.errorTotal
+	ch <- c.latency
+	ch <- c.kubernetesErrorTotal
+}
+
+func (c *NetworkCollector) Collect(ch chan<- prometheus.Metric) {
+	hosts := []string{c.host}
+
+	endpoints, err := c.kubernetesClient.CoreV1().Endpoints(c.namespace).Get(c.serviceName, metav1.GetOptions{})
+	if err != nil {
+		c.logger.Log("level", "error", "message", "could not get endpoints", "stack", fmt.Sprintf("%#v", err))
+		c.kubernetesErrorCount++
+		ch <- prometheus.MustNewConstMetric(c.kubernetesErrorTotal, prometheus.CounterValue, c.kubernetesErrorCount)
+
+		return
+	}
+
+	for _, endpointSubset := range endpoints.Subsets {
+		for _, address := range endpointSubset.Addresses {
+			hosts = append(hosts, fmt.Sprintf("%v:%v", address.IP, c.port))
+		}
+	}
+
+	var wg sync.WaitGroup
+
+	for _, host := range hosts {
+		wg.Add(1)
+
+		go func(host string) {
+			defer wg.Done()
+
+			start := time.Now()
+
+			c.countMutex.Lock()
+			c.count[host]++
```

Comment:
Should this be increased in case of error?

Reply:
I think so. We attempted to make a check, so the count increases, and it failed, so the error count increases. Makes it easier to work out a percentage of failures.